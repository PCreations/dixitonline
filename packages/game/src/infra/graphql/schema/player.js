import { objectType } from 'nexus';

export const Player = objectType({
  name: 'GamePlayer',
  definition(t) {
    t.id('id');
    t.string('name');
    t.int('score', {
      resolve(player) {
        return player.score ?? 0;
      },
    });
  },
});
