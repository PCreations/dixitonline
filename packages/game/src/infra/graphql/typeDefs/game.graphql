type Game {
  id: ID!
  host: GamePlayer!
  players: [GamePlayer!]!
}
