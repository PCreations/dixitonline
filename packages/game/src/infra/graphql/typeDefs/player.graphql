type GamePlayer {
  id: ID!
  name: String!
}
