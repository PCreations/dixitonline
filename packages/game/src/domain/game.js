/* eslint-disable max-classes-per-file */
import {
  newGameCreatedEvent,
  playerJoinedGame,
  newGameStartedEvent,
  handsCompletedEvent,
  gameEndedEvent,
} from './events';
import { equals as playerEquals } from './player';
import { makeResult, makeErrorResult } from './result';

export const DEFAULT_END_CONDITION = {};

export const GameStatus = {
  WAITING_FOR_PLAYERS: 'WAITING_FOR_PLAYERS',
  STARTED: 'STARTED',
  ENDED: 'ENDED',
};

export const GameError = {
  GAME_ALREADY_JOINED: 'GAME_ALREADY_JOINED',
  GAME_ALREADY_STARTED: 'GAME_ALREADY_STARTED',
  MAXIMUM_NUMBER_OF_PLAYERS_REACHED: 'MAXIMUM_NUMBER_OF_PLAYERS_REACHED',
  ONLY_HOST_CAN_START_GAME: 'ONLY_HOST_CAN_START_GAME',
  NOT_ENOUGH_PLAYERS: 'NOT_ENOUGH_PLAYERS',
};

export const MAXIMUM_NUMBER_OF_PLAYERS = 6;
export const MINIMUM_NUMBER_OF_PLAYERS = 3;
export const NUMBER_OF_CARDS_IN_A_DECK = 84;

export const makeGame = ({
  id,
  host,
  cards = [],
  score = {},
  players = [],
  status = GameStatus.WAITING_FOR_PLAYERS,
  currentTurn = { id: null, storytellerId: null, number: 0 },
  endCondition = DEFAULT_END_CONDITION,
} = {}) => {
  if (!id) throw new Error('Game must contain an id');
  if (!host) throw new Error('Game must have an host');
  if (
    endCondition !== DEFAULT_END_CONDITION &&
    typeof endCondition.xTimesStorytellerLimit === 'undefined' &&
    typeof endCondition.scoreLimit === 'undefined'
  )
    throw new Error(`Invalid end condition, received "${endCondition}"`);

  console.log(`building a game with ${cards.length} cards`);

  return Object.freeze({
    id,
    host,
    players,
    cards,
    endCondition,
    score,
    status,
    currentTurn,
  });
};

export const getAllPlayers = game => [game.host, ...game.players];

export const getEndCondition = game => {
  const allPlayersLength = getAllPlayers(game).length;
  if (typeof game.endCondition.xTimesStorytellerLimit !== 'undefined') {
    return {
      remainingTurns: allPlayersLength * game.endCondition.xTimesStorytellerLimit - game.currentTurn.number,
    };
  }
  if (typeof game.endCondition.scoreLimit !== 'undefined') {
    return {
      scoreLimit: game.endCondition.scoreLimit,
    };
  }
  return {
    remainingTurns: Math.floor(game.cards.length / allPlayersLength),
  };
};

const isGameFull = game => getAllPlayers(game).length === MAXIMUM_NUMBER_OF_PLAYERS;

export const createGame = ({ gameId, host, endCondition }) => {
  return makeResult(makeGame({ id: gameId, host, endCondition }), [newGameCreatedEvent({ gameId })]);
};

export const joinPlayer = (game, player) => {
  if (getAllPlayers(game).some(playerEquals.bind(null, player))) return makeErrorResult(GameError.GAME_ALREADY_JOINED);
  if (isGameFull(game)) return makeErrorResult(GameError.MAXIMUM_NUMBER_OF_PLAYERS_REACHED);
  return makeResult(makeGame({ ...game, players: game.players.concat(player) }), [
    playerJoinedGame({ gameId: game.id, playerId: player.id }),
  ]);
};

export const startGame = (game, player) => {
  if (game.status === GameStatus.STARTED) {
    return makeErrorResult(GameError.GAME_ALREADY_STARTED);
  }
  if (getAllPlayers(game).length < MINIMUM_NUMBER_OF_PLAYERS) {
    return makeErrorResult(GameError.NOT_ENOUGH_PLAYERS);
  }
  if (playerEquals(game.host, player)) {
    return makeResult(
      makeGame({
        ...game,
        status: GameStatus.STARTED,
      }),
      [newGameStartedEvent({ gameId: game.id, playerIds: getAllPlayers(game).map(({ id }) => id) })]
    );
  }
  return makeErrorResult(GameError.ONLY_HOST_CAN_START_GAME);
};

export const completeHands = (game, { cards, actualHandsByPlayerId, previousTurnId }) => {
  const allPlayers = getAllPlayers(game);
  const actualCards = cards ?? game.cards;
  console.log('actualCards length', actualCards.length);
  if (!actualHandsByPlayerId) {
    const handsByPlayerId = allPlayers.reduce(
      (hands, player, playerIndex) => ({
        ...hands,
        [player.id]: actualCards.slice(playerIndex * 6, playerIndex * 6 + 6),
      }),
      {}
    );
    const remainingCards = actualCards.slice(allPlayers.length * 6);
    return makeResult(
      makeGame({
        ...game,
        cards: remainingCards,
      }),
      [handsCompletedEvent({ gameId: game.id, handsByPlayerId })]
    );
  }
  if (actualCards.length < allPlayers.length) {
    return makeResult(
      makeGame({
        ...game,
        status: GameStatus.ENDED,
      }),
      [gameEndedEvent({ gameId: game.id })]
    );
  }
  const handsByPlayerId = Object.entries(actualHandsByPlayerId).reduce(
    (completedHands, [playerId, hand], playerIndex) => ({
      ...completedHands,
      [playerId]: hand.concat(actualCards[playerIndex]),
    }),
    {}
  );
  console.log('new actual cards length', actualCards.slice(allPlayers.length).length);
  return makeResult(
    makeGame({
      ...game,
      cards: actualCards.slice(allPlayers.length),
    }),
    [handsCompletedEvent({ gameId: game.id, handsByPlayerId, previousTurnId })]
  );
};

export const updateScore = (game, turnScore) => {
  const newScore = Object.entries(turnScore).reduce(
    (scores, [playerId, playerTurnScore]) => ({
      ...scores,
      [playerId]: playerTurnScore + (game.score[playerId] || 0),
    }),
    {}
  );
  console.log('updated score', newScore);
  return makeResult(
    makeGame({
      ...game,
      score: newScore,
    })
  );
};

export const setCurrentTurn = (game, currentTurn) => makeResult(makeGame({ ...game, currentTurn }));
