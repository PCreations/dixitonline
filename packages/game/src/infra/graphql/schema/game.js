import { objectType, enumType, unionType } from 'nexus';
import { Player } from './player';
import { getAllPlayers, getEndCondition, GameStatus as DomainGameStatus } from '../../../domain/game';

export const GameStatus = enumType({
  name: 'GameStatus',
  members: DomainGameStatus,
});

export const RemainingTurnsEndCondition = objectType({
  name: 'GameRemainingTurnsEndCondition',
  definition(t) {
    t.int('remainingTurns');
  },
});

export const ScoreLimitEndCondition = objectType({
  name: 'GameScoreLimitEndCondition',
  definition(t) {
    t.int('scoreLimit');
  },
});

export const GameEndCondition = unionType({
  name: 'GameEndCondition',
  definition(t) {
    t.members(RemainingTurnsEndCondition, ScoreLimitEndCondition);
    t.resolveType(obj =>
      typeof obj.remainingTurns === 'undefined' ? 'GameScoreLimitEndCondition' : 'GameRemainingTurnsEndCondition'
    );
  },
});

export const Game = objectType({
  name: 'Game',
  definition(t) {
    t.id('id');
    t.field('host', { type: Player });
    t.field('status', { type: GameStatus });
    t.id('currentTurnId', {
      nullable: true,
      resolve({ currentTurn }) {
        return currentTurn.id;
      },
    });
    t.int('remainingTurns', {
      deprecation: 'use endCondition instead',
    });
    t.field('endCondition', {
      type: GameEndCondition,
      resolve(game) {
        console.log(getEndCondition(game));
        return getEndCondition(game);
      },
    });
    t.list.field('players', {
      type: Player,
      resolve(game) {
        const { score } = game;
        const players = getAllPlayers(game).map(player => {
          return {
            ...player,
            score: score[player.id],
          };
        });
        return players;
      },
    });
  },
});
