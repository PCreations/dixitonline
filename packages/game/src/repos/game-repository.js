import shortid from 'shortid';
import { makeGame, GameStatus } from '../domain/game';

export class GameNotFoundError extends Error {
  constructor(gameId) {
    super();
    this.message = `Game ${gameId} not found`;
  }
}

export const makeGameRepository = ({ uuid = shortid, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    saveGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      if (!doc.exists) {
        throw new GameNotFoundError(id);
      }
      return makeGame(doc.data());
    },
    getAllGames() {
      return lobbyGames
        .where('status', '==', GameStatus.WAITING_FOR_PLAYERS)
        .get()
        .then(snapshot => {
          const games = [];
          snapshot.forEach(doc => games.push(makeGame(doc.data())));
          return games;
        });
    },
    deleteGameById(gameId) {
      return lobbyGames.doc(gameId).delete();
    },
  };
};

export const makeNullGameRepository = ({ nextGameId, gamesData }) =>
  makeGameRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        where(field, _, value) {
          return {
            async get() {
              const docs = Object.values(gamesData)
                .filter(Boolean)
                .filter(game => game[field] === value)
                .map(game => ({
                  data() {
                    return game;
                  },
                }));
              return {
                forEach: docs.forEach.bind(docs),
              };
            },
          };
        },
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
                exists: typeof gamesData[gameId] !== 'undefined',
              };
            },
            async set(game) {
              gamesData[gameId] = game;
            },
            delete() {
              gamesData[gameId] = undefined;
            },
          };
        },
      };
    },
  };
};
