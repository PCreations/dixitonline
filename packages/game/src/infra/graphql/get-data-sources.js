import { makeGameRepository } from '../../repos/game-repository';

export const makeGetDataSources = ({ gameRepository = makeGameRepository() } = {}) => () => ({
  gameRepository,
});
