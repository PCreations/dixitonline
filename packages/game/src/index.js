import * as GameTypes from './infra/graphql/schema';
import * as events from './domain/events';
import { makeGameRepository } from './repos/game-repository';
import { initialize as initializeSubscribers } from './subscribers';
import { makeGetDataSources } from './infra/graphql/get-data-sources';
import { makeGetContext } from './infra/graphql/get-context';

export const initialize = ({ firestore, authorizationService, dispatchDomainEvents, subscribeToDomainEvent }) => {
  const gameRepository = makeGameRepository({ firestore });
  const getDataSources = makeGetDataSources({ gameRepository });
  const getContext = makeGetContext({ dispatchDomainEvents, authorizationService });
  initializeSubscribers({ subscribeToDomainEvent, dispatchDomainEvents, gameRepository });

  return {
    getDataSources,
    getContext,
  };
};
export { GameTypes };
export { events };
