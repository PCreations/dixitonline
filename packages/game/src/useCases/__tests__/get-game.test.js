import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import { makeNullGameRepository } from '../../repos/game-repository';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';
import { makeTestServer } from '../../__tests__/test-server';
import { buildTestGame } from '../../__tests__/dataBuilders/game';
import { buildgameRepositoryInitialGames } from '../../__tests__/dataBuilders/game-repository-initial-games';
import { buildTestPlayer } from '../../__tests__/dataBuilders/player';

describe('getGame', () => {
  test('gets a game by its id', async () => {
    // arrange
    const host = buildTestPlayer()
      .withId('p1')
      .build();
    const players = [
      buildTestPlayer()
        .withId('p2')
        .build(),
      buildTestPlayer()
        .withId('p3')
        .build(),
      buildTestPlayer()
        .withId('p4')
        .build(),
      buildTestPlayer()
        .withId('p5')
        .build(),
    ];
    const initialGames = buildgameRepositoryInitialGames()
      .withGames([
        buildTestGame()
          .withId('g1')
          .withCards(54)
          .withHost(host)
          .withCurrentTurnId('t1')
          .withPlayers(players)
          .withScore([2, 4, 6, 9, 3])
          .build(),
      ])
      .build();
    const server = makeTestServer({
      getDataSources: makeGetDataSources({
        gameRepository: makeNullGameRepository({
          nextGameId: 'g2',
          gamesData: initialGames,
        }),
      }),
      currentUserId: 'p1',
      currentUserUsername: 'player1',
    });
    const GET_GAME = gql`
      query GetGame($gameId: ID!) {
        game(gameId: $gameId) {
          id
          currentTurnId
          endCondition {
            __typename
            ... on GameRemainingTurnsEndCondition {
              remainingTurns
            }
          }
          status
          players {
            score
          }
        }
      }
    `;

    // act
    const { query } = createTestClient(server);

    // assert
    const response = await query({ query: GET_GAME, variables: { gameId: 'g1' }, operationName: 'GetGame' });
    expect(response).toMatchSnapshot();
  });
});
