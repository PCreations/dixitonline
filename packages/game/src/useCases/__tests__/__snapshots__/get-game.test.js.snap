// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGame gets a game by its id 1`] = `
Object {
  "data": Object {
    "game": Object {
      "currentTurnId": "t1",
      "endCondition": Object {
        "__typename": "GameRemainingTurnsEndCondition",
        "remainingTurns": 10,
      },
      "id": "g1",
      "players": Array [
        Object {
          "score": 2,
        },
        Object {
          "score": 4,
        },
        Object {
          "score": 6,
        },
        Object {
          "score": 9,
        },
        Object {
          "score": 3,
        },
      ],
      "status": "WAITING_FOR_PLAYERS",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
